sql/update.go

```diff
@@ -18,11 +18,105 @@
 package sql
 
 import (
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/client"
 	"github.com/cockroachdb/cockroach/sql/parser"
-	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/log"
 )
 
-// Update TODO(pmattis): document.
+// Update updates columns for a selection of rows from a table.
+// TODO(vivek): Update columns that are a part of an index.
 func (p *planner) Update(n *parser.Update) (planNode, error) {
-	return nil, util.Errorf("TODO(pmattis): unimplemented")
+	tableDesc, err := p.getAliasedTableDesc(n.Table)
+	if err != nil {
+		return nil, err
+	}
+
+	// Determine which columns we're inserting into.
+	var names parser.QualifiedNames
+	for _, expr := range n.Exprs {
+		names = append(names, expr.Name)
+	}
+	cols, err := p.processColumns(tableDesc, names)
+	if err != nil {
+		return nil, err
+	}
+
+	// Don't allow updating any column that is part of the primary key.
+	updateColMap := map[uint32]struct{}{}
+	for _, c := range cols {
+		updateColMap[c.ID] = struct{}{}
+	}
+	for i, id := range tableDesc.Indexes[0].ColumnIDs {
+		if _, ok := updateColMap[id]; ok {
+			return nil, fmt.Errorf("disallow updating primary key column \"%s\"", tableDesc.Indexes[0].ColumnNames[i])
+		}
+	}
+
+	// Query the rows that need updating.
+	// TODO(vivek): Avoid going through Select.
+	row, err := p.Select(&parser.Select{
+		Exprs: parser.SelectExprs{
+			&parser.StarExpr{TableName: parser.Name(tableDesc.Name)},
+		},
+		From:  parser.TableExprs{n.Table},
+		Where: n.Where,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a map of all the columns.
+	colMap := map[uint32]int{}
+	for i, name := range row.Columns() {
+		c, err := tableDesc.FindColumnByName(name)
+		if err != nil {
+			return nil, err
+		}
+		colMap[c.ID] = i
+	}
+
+	index := tableDesc.Indexes[0]
+	indexKey := encodeIndexKeyPrefix(tableDesc.ID, index.ID)
+
+	// Evaluate all the column value expressions.
+	vals := make([]parser.Expr, 0, 10)
+	for _, expr := range n.Exprs {
+		val, err := parser.EvalExpr(expr.Expr, nil)
+		if err != nil {
+			return nil, err
+		}
+		vals = append(vals, val)
+	}
+
+	// Update all the rows.
+	b := client.Batch{}
+	for row.Next() {
+		if err := row.Err(); err != nil {
+			return nil, err
+		}
+		// TODO(tamird/pmattis): update the secondary indexes too.
+		primaryKey, err := encodeIndexKey(index, colMap, row.Values(), indexKey)
+		if err != nil {
+			return nil, err
+		}
+		// Update the columns.
+		for i, val := range vals {
+			key := encodeColumnKey(cols[i], primaryKey)
```

Comment:
That's a good question. I think technically the transaction should go through only if the rows picked with the WHERE clause are the ones intended to be picked by the transaction. That is enforced through the transaction and doesn't need CPut.

In particular, if we had to apply CPut we would also need to run CPut with all the other columns that are used in the WHERE clause.